### PR TITLE
feat: stricter parent types for the AST

### LINKED ISSUE #6225: Enhancement: add strict `parent` types for nodes that have well-defined parents
Right now a node's `.parent` property is loosely typed as `TESTree.Node`. This is pretty sucky as it means that users need to either add unnecessary logic or explicit casts to refine types in some cases.

There are a number of nodes wherein we know exactly what the parent node will always be due to the defined structure of the AST.
For example, we know that a `VariableDeclarator` will always have a `VariableDeclaration` parent.

We should define some of these simpler relationships in our types to improve the experience of consuming the AST types.

Some cases we can define:
- [x] `AccessorProperty` has parent `ClassBody` (#9487)
- [ ] `CatchClause` has parent `TryStatement`
- [ ] `ClassBody` has parent `ClassDeclaration | ClassExpression | TSAbstractClassDeclaration`
- [ ] `ExportSpecifier` has parent `ExportNamedDeclaration`
- [ ] `ImportAttribute` has parent `ImportDeclaration | ImportExpression`
- [ ] `ImportDefaultSpecifier` has parent `ImportDeclaration`
- [ ] `ImportNamespaceSpecifier` has parent `ImportDeclaration`
- [ ] `ImportSpecifier` has parent `ExportAllDeclaration | ExportNamedDeclaration | ImportDeclaration`
- [ ] `JSXAttribute` has parent `JSXOpeningElement`
- [ ] `JSXClosingElement` has parent `JSXElement`
- [ ] `JSXClosingFragment` has parent `JSXFragment`
- [ ] `JSXOpeningElement` has parent `JSXElement`
- [ ] `JSXOpeningFragment` has parent `JSXFragment`
- [ ] `JSXSpreadAttribute` has parent `JSXOpeningElement`
- [ ] `MethodDefinition` has parent `ClassBody`
- [ ] `Property` has parent `ObjectExpression | ObjectPattern`
- [ ] `PropertyDefinition` has parent `ClassBody`
- [ ] `SpreadElement` has parent `ArrayExpression | CallExpression | ObjectExpression`
- [ ] `StaticBlock` has parent `ClassBody`
- [ ] `SwitchCase` has parent `SwitchStatement`
- [ ] `TemplateElement` has parent `TemplateLiteral | TSTemplateLiteralType`
- [ ] `TSAbstractAccessorProperty` has parent `ClassBody`
- [ ] `TSAbstractMethodDefinition` has parent `ClassBody`
- [ ] `TSAbstractPropertyDefinition` has parent `ClassBody`
- [ ] `TSCallSignatureDeclaration` has parent `TSInterfaceBody | TSTypeLiteral`
- [ ] `TSConstructSignatureDeclaration` has parent `TSInterfaceBody | TSTypeLiteral`
- [ ] `TSClassImplements` has parent `ClassDeclaration | ClassExpression`
- [ ] `TSEnumMember` has parent `TSEnumDeclaration`
- [ ] `TSIndexSignature` has parent `ClassBody | TSInterfaceBody | TSTypeLiteral`
- [ ] `TSInterfaceBody` has parent `TSInterfaceDeclaration`
- [ ] `TSInterfaceHeritage` has parent `TSInterfaceBody`
- [ ] `TSMethodSignature` has parent `TSInterfaceBody | TSTypeLiteral`
- [ ] `TSModuleBlock` has parent `TSModuleDeclaration`
- [ ] `TSParameterProperty` has parent `FunctionLike`
- [ ] `TSPropertySignature` has parent `TSInterfaceBody | TSTypeLiteral`
- [ ] `TSTypeParameter` has parent `TSInferType | TSTypeParameterDeclaration | TSMappedType`
- [ ] `VariableDeclarator` has parent `VariableDeclaration` (#9228)

Note: the implementor should double check that I've written these correctly by comparing them against the AST types.

The import thing to note is that we shouldn't define all of the relationships that exist - just a subset of them that we can easily statically define and maintain. I'd probably put a rule around ~3 parent types being the maximum.

### PR DESCRIPTION
**Pull request by bradzacher**
## PR Checklist

- [x] Addresses an existing open issue: fixes #6225
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [Contributing](https://typescript-eslint.io/contributing) were taken

## Overview

Adds all the parent types specified in #6225 and fixes the resulting type and linting errors.
Pretty straightforward here!

### CODE REVIEW

**Review by kirkwaiblinger — APPROVED** [review]
smart

File: packages/types/src/ts-estree.ts [hunk 1]
```diff
@@ -16,10 +16,189 @@ declare module './generated/ast-spec' {
...
+
+  interface ImportNamespaceSpecifier {
+    parent: TSESTree.ImportDeclaration;
+  }
+
+  interface ImportSpecifier {
+    parent:
+      | TSESTree.ExportAllDeclaration
+      | ExportNamedDeclaration
+      | ImportDeclaration;
```
Comment (bradzacher) [inline packages/types/src/ts-estree.ts:57]:
```suggestion
      | TSESTree.ExportNamedDeclaration
      | TSESTree.ImportDeclaration;
```

File: packages/types/src/ts-estree.ts [hunk 2]
```diff
@@ -16,10 +16,189 @@ declare module './generated/ast-spec' {
...
+  interface ClassBody {
+    parent: TSESTree.ClassExpression | TSESTree.ClassDeclaration;
+  }
+
+  interface ExportSpecifier {
+    parent: TSESTree.ExportNamedDeclaration;
+  }
+
+  interface ImportAttribute {
+    parent: TSESTree.ImportDeclaration | ImportExpression;
```
Comment (bradzacher) [inline packages/types/src/ts-estree.ts:42]:
```suggestion
    parent: TSESTree.ImportDeclaration | TSESTree.ImportExpression;
```

**Review by Josh-Cena — APPROVED** [review]
Do you know how we can keep this exhaustive into the future, or is it just do-it-as-you-please?

**Comment by bradzacher** [comment]
I think that it's probably going to be very difficult to keep it exhaustive.
We can probably write a script which scans the AST declaration to find the best parent lists though?

Once we've declared it to a great extent once it should be trivial to keep it updated with any future AST changes.

**Pull request merged**